be/src/util/string_parser.hpp

```diff
@@ -429,114 +418,95 @@ inline T StringParser::string_to_float_internal(const char* s, int len, ParseRes
                 // here and interpret any column with inf as a prefix as infinity rather than
                 // checking every remaining byte.
                 *result = PARSE_SUCCESS;
                 return negative ? -INFINITY : INFINITY;
             } else {
-                // Starts with 'i', but isn't inf...
+                // Starts with 'i   ', but isn't inf...
                 *result = PARSE_FAILURE;
                 return 0;
             }
-        } else if (s[i] == 'n' || s[i] == 'N') {
-            if (len > i + 2 && (s[i + 1] == 'a' || s[i + 1] == 'A') &&
-                (s[i + 2] == 'n' || s[i + 2] == 'N')) {
+        } else if (s[k] == 'n' || s[k] == 'N') {
+            if (len > k + 2 && (s[k + 1] == 'a' || s[k + 1] == 'A') &&
+                (s[k + 2] == 'n' || s[k + 2] == 'N')) {
                 *result = PARSE_SUCCESS;
                 return negative ? -NAN : NAN;
             } else {
                 // Starts with 'n', but isn't NaN...
                 *result = PARSE_FAILURE;
                 return 0;
             }
-        } else {
-            if ((UNLIKELY(i == first || !is_all_whitespace(s + i, len - i)))) {
-                // Reject the string because either the first char was not a digit, "," or "e",
-                // or the remaining chars are not all whitespace
-                *result = PARSE_FAILURE;
-                return 0;
-            }
-            // skip trailing whitespace.
-            break;
         }
     }
 
-    val += remainder / divide;
-
-    if (i < len && (s[i] == 'e' || s[i] == 'E')) {
-        // Create a C-string from s starting after the optional '-' sign and fall back to
-        // strtod to avoid conversion inaccuracy for scientific notation.
-        // Do not use boost::lexical_cast because it causes codegen to crash for an
-        // unknown reason (exception handling?).
-        char c_str[len - negative + 1];
-        memcpy(c_str, s + negative, len - negative);
-        c_str[len - negative] = '\0';
-        char* s_end;
-        val = strtod(c_str, &s_end);
-        if (s_end != c_str + len - negative) {
-            // skip trailing whitespace
-            int trailing_len = len - negative - (int)(s_end - c_str);
-            if (UNLIKELY(!is_all_whitespace(s_end, trailing_len))) {
-                *result = PARSE_FAILURE;
-                return val;
-            }
+    // check invalid char
+    for (int k = i; k <= j; ++k) {
+        if ((s[k] >= '0' && s[k] <= '9') || s[k] == '.' || s[k] == 'e' || s[k] == 'E' || s[k] == '-' || s[k] == '+') {
+            continue;
+        } else {
+            *result = PARSE_FAILURE;
+            return 0;
         }
     }
 
-    // Determine if it is an overflow case and update the result
-    if (UNLIKELY(val == std::numeric_limits<T>::infinity())) {
-        *result = PARSE_OVERFLOW;
-    } else {
-        *result = PARSE_SUCCESS;
+    double val;
+    auto res = fast_float::from_chars(s + i, s + j + 1, val);
```

Comment:
If you import the fast_float library, is there necessary to change the string_parser.hpp?

Reply:
i think it is necessary, original string_parser.hpp use heristic method to convert string to double(val = val + remainder / divide), it will lose some precision, we need to discard this implementation.